rs/rust_canisters/xnet_test/src/main.rs

```diff
@@ -213,85 +191,76 @@ fn fanout() {
-                ));
-                METRICS.with(|m| m.borrow_mut().call_errors += 1);
-            } else {
-                METRICS.with(move |m| m.borrow_mut().requests_sent += 1);
+                (payload,),
+            )
+            .await;
+
+            match res {
+                Ok((reply,)) => {
+                    let elapsed = Duration::from_nanos(time_nanos() - reply.time_nanos);
+                    METRICS.with(move |m| m.borrow_mut().requests_sent += 1);
+                    METRICS.with(|m| m.borrow_mut().latency_distribution.observe(elapsed));
+                }
+                Err((err_code, _)) => {
+                    log(&format!(
+                        "{} call failed with {:?}",
+                        time_nanos() / 1_000_000,
+                        err_code
+                    ));
+                    METRICS.with(|m| m.borrow_mut().call_errors += 1);
+                    METRICS.with(|m| m.borrow_mut().reject_responses += 1);
+                }
             }
         }
     }
 }
 
 /// Endpoint that handles requests from canisters located on remote subnets.
-#[export_name = "canister_update handle_request"]
-fn handle_request() {
-    let (req, _) =
-        candid::Decode!(&api::arg_data()[..], Request, Vec<u8>).expect("failed to decode request");
-    let caller = api::caller();
-    let in_seq_no = STATE.with(|s| s.borrow_mut().set_in_seq_no(caller.into_vec(), req.seq_no));
+#[update]
+fn handle_request(req: Request) -> Reply {
+    let caller = caller();
+    let in_seq_no = STATE.with(|s| s.borrow_mut().set_in_seq_no(caller, req.seq_no));
 
     if req.seq_no <= in_seq_no {
         METRICS.with(|m| m.borrow_mut().seq_errors += 1);
     }
 
-    let msg = candid_encode_padded(&Reply {
+    Reply {
         time_nanos: req.time_nanos,
-    });
-    api::reply(&msg[..]);
+    }
 }
 
 /// Deposits the cycles this canister has minus 1T according to the given
 /// `DepositCyclesArgs`
-#[export_name = "canister_update return_cycles"]
-fn return_cycles() {
-    let cycle_refund = api::canister_cycle_balance().saturating_sub(1_000_000_000_000);
-    let noop = |_| ();
-    let _ = api::call_raw(
-        api::CanisterId::from_str("aaaaa-aa").unwrap(),
+#[update]
+async fn return_cycles(canister_id_record: CanisterIdRecord) -> String {
+    let cycle_refund = canister_balance().saturating_sub(1_000_000_000_000);
+    let _ = call_with_payment::<(CanisterIdRecord,), ()>(
+        Principal::from_str("aaaaa-aa").unwrap(),
         "deposit_cycles",
-        &api::arg_data(),
-        noop,
-        noop,
-        None,
-        std::ptr::null_mut(),
-        api::Funds {
-            cycles: cycle_refund,
-        },
-    );
-
-    candid_reply(&"ok");
+        (canister_id_record,),
+        cycle_refund,
+    )
+    .await;
+
+    "ok".to_string()
 }
 
 /// Query call that serializes metrics as a candid message.
-#[export_name = "canister_query metrics"]
-fn metrics() {
-    let msg = METRICS
```

Comment:
The panic line `rs/messaging/tests/queue_tests.rs:503:73` calls this method with an empty vector. Note that the original implementation doesn't even do decoding for the input. When changing to `#[query]`, you cannot skip decoding, so you need to send in `Encode!()`, instead of `Vec::new()`.